### ISSUE TITLE
Writing in constant expression to inactive union member via reference results in error

### ISSUE THREAD

**Issue by Fedr** [llvm/llvm-project#62988]
The following program:
```
constexpr int f()
{
    union {
        int x{0};
        char y;
    };
    auto & z = y;
    z = 1;
    x = 0;
    return x;
}

static_assert( f() == 0 );
```
is accepted by GCC and MSVC. But Clang complains:
```
assignment to member 'y' of union with active member 'x' is not allowed in a constant expression
```
Online demo: https://gcc.godbolt.org/z/EsnEsTKP4

**Comment by shafik** [comment]
This is a C++17 restriction, if we set gcc to C++17 then it also fails: https://gcc.godbolt.org/z/esGnbPdba

See C++17 [expr.const p2.10](https://timsong-cpp.github.io/cppwp/n4659/expr.const#2.10):

> an assignment expression or invocation of an assignment operator ([class.copy]) that would change the active member of a union;

This changed w/ C++20 via [p0137r1](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0137r1.html)

**Comment by frederick-vs-ja** [comment]
If I understood correctly, [P0137R1](https://wg21.link/p0137r1) only allows assignment to change active member if the assignment is performed on restricted forms where the enclosing union must be known ([[class.union.general]/5](https://eel.is/c++draft/class.union.general#5)). So `z = 1` can't change active member of the union and thus results in UB. It's probably intended that `z = 1` is not equivalent to `y = 1` here.

I think Clang has been doing the right thing.

**Comment by zygoloid** [comment]
Yes, the intent is that `y = 1;` is a valid way to change the active union member, and `z = 1;` is not.
